### ISSUE TITLE
Scraping misses series periodically

### ISSUE THREAD

**Issue by mamoep** [VictoriaMetrics/VictoriaMetrics#1857]
**Describe the bug**
I am using VictoriaMetrics single node to scrape targets of the software "NetApp Harvest". Periodically the number of metrics scraped is lower then the number of metrics emitted. The time pattern matches the interval at which Harvest emits the metrics.

**To Reproduce**
Setup NetApp Harvest 21.11 (https://github.com/NetApp/harvest) to scrape NetApp storage system. (I can provide details if necessary)
Setup VictoriaMetrics with the following options

config.yml content:
```yaml
scrape_configs:
- http_sd_configs:
  - url: http://localhost:8887/api/v1/sd
  job_name: harvest
  scrape_interval: 15s
```

Query: `count(net_port_status{cluster="NetAppClusterName"})`

**Expected behavior**
Harvest emtis all network ports of the NetApp system, so count(net_port_status) should give a flat line with the same number for every scrape interval.

**Logs**
No warnings or errors visible

**Screenshots**
VictoriaMetrics vs. Prometheus
![Screenshot 2021-11-24 165923](https://user-images.githubusercontent.com/26142342/143273040-7809bdc3-6243-414e-9146-23348f8010c6.jpg)

**Version**
```
$ ./victoria-metrics-prod --version
victoria-metrics-20211108-134932-tags-v1.69.0-0-g69e655ba7
```

**Used command-line flags**
```
 --promscrape.maxScrapeSize=100MB
 --promscrape.config.strictParse
 --snapshotAuthKey=...
 --retentionPeriod=6
 --storageDataPath=/data/victoriametrics
 --selfScrapeInterval=10s
 --deleteAuthKey=...
 --graphiteListenAddr=:2003
 --promscrape.config=/etc/victoriametrics/config.yml
```

**Comment by mamoep** [comment]
I just waded through the previous releases. The issue is introduced with v1.66.0. 

Versions 1.62.0 to 1.65.0 are not affected.

**Comment by mamoep** [comment]
Further testing shows that setting "-promscrape.noStaleMarkers" resolves the issue. I am using a Single Node without any separate vmagent or remote write.

**Comment by valyala** [comment]
If the `-promscrape.noStaleMarkers` command-line flag resolves the issue, then it is likely the issue is related to [staleness handling](https://docs.victoriametrics.com/vmagent.html#prometheus-staleness-markers). Probably the `NetApp Harvest` exporter periodically stops exporting some metrics, which results in gaps on graphs. Another possible issue may be related to the changed handling of timestamps since v1.68.0 - see https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1807#issuecomment-984523424 for details.

**Comment by mamoep** [comment]
I also thought about the staleness handling. If it was an issue with of the exporter, shouldn't the effect show in native Prometheus as well? For testing I "curl"ed the exporter every 1 second and always got the same number of metrics. Do you have any better ideas to debug this?
My tests indicate the issue started with v1.66.0 so I don't think it is related to timestamp handling change in v1.68.0.

**Comment by mamoep** [comment]
The counter `scrape_samples_scraped{job="<job-name>", instance="<instance-id>"} `shows a constant number over time.

**Comment by mamoep** [comment]
I did a tcpdump now and I can exclude any missing metrics from the Harvest exporter. The order of the metrics page varies a lot. The exporter always pushes the new metrics to the top of the page. Could this be the source of the issue? I would appreciate any suggestion how to do further debugging.

**Comment by mamoep** [comment]
I found that the Harvest exporter is not sorting the labels.

I tried to understand what's going on on the code level in VM. If I understand it correctly `parser.GetRowsDiff `doesn't account for changed order of labels and so `sendStaleSeries` marks everything stale where the label order changed. The command line flag `sortLabels` is only taken into account for the write not for the stale check.

Can you confirm my analysis?

**Comment by 7840vz** [comment]
Can confirm same issue with harvest and [idrac_exporter](https://github.com/mrlhansen/idrac_exporter). Both exporters don't sort labels but returns same series otherwise:

So first scrape returns:

```
metric1{label_a="x", label_b="y"} 1
```

Second scrape of the same host returns:

```
metric1{label_b="y", label_a="x"} 1
```

Looks like this cause vmagent to mark `metric1` as stale.

**Comment by hagen1778** [comment]
Indeed, for staleness detection vmagent compares bodies of the current and previous scrapes. Lines are compared as bytes, ignoring only values. So labels in different order will be accounted as a mismatch. The sorting is likely omitted due to performance reasons. I'll see if this can be achieved without significant performance impact.

**Comment by hagen1778** [comment]
So far there is no easy way to re-use `sortLabels`  cmd-line flag, since it is dedicated to remoteWrite package. Using it for both scraping and remote-writing may complicate the code without bringing many benefits. Adding new cmd-line flag for sorting labels on scrape seems like overkill.
Let's keep the issue open for now until a better solution is found.

**Comment by valyala** [comment]
@hagen1778 , how about adding `sort_labels` option at [scrape_config](https://docs.victoriametrics.com/sd_configs.html#scrape_configs) level? If it is set to true, then sort labels in scraped metrics before checking for [stale metrics](https://docs.victoriametrics.com/vmagent.html#prometheus-staleness-markers). By default the `sort_labels` option must be set to `false` in order to avoid performance slowdown related to labels' sorting. Then users could set `sort_labels` to `true` for scrape targets, which may change the order of labels per each exported metric, in order to properly handle staleness detection at the cost of some performance degradation.

**Comment by hagen1778** [comment]
> how about adding sort_labels option at [scrape_config](https://docs.victoriametrics.com/sd_configs.html#scrape_configs) level?

That's one of the options, yes. What makes me feel weird about it is that this sorting would matter only for staleness procedure. For everything else, it doesn't matter.

**Comment by mamoep** [comment]
From my point of you this shouldn't be an option of any kind. You claim to be a drop-in replacement for Prometheus and in this case it just fails. This is a bug and should be solved "behind the curtain".

**Comment by valyala** [comment]
The temporary workaround is to disable [staleness markers](https://docs.victoriametrics.com/vmagent.html#prometheus-staleness-markers) for scrape targets, which change the order of labels for the same metrics between scrapes. This can be done in the following ways:

- By specifying `no_stale_markers` option at the [scrape_config](https://docs.victoriametrics.com/sd_configs/#scrape_configs) for the given target. In this case staleness markers for other scrape targets remain enabled.
- By running `vmagent` with `-promscrape.noStaleMarkers` - in this case staleness markers for all scrape targets are disabled.
- By passing `sort_labels: true` to Harvest as suggested [above](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1857#issuecomment-1477771496).

Side note: all the VictoriaMetrics components assume the order of labels for the same [time series](https://docs.victoriametrics.com/keyconcepts/#time-series) do not change across the collected [raw samples](https://docs.victoriametrics.com/keyconcepts/#raw-samples). This allows avoiding expensive sorting of labels for the ingested samples and improving data ingestion performance. VictoriaMetrics itself properly accepts samples with changing order of labels, but in this case it falls back to slow path, which requires higher CPU and RAM usage. That's why it is recommended keeping the order of labels the same across the ingested samples.

**Comment by mamoep** [comment]
I fully understand your concern for performance. But I have two problems with your comment:
1. Claiming that VictoriaMetrics accepts samples with changing label order is only partly true. The broken staleness handling will lead to dropped samples, as documented above.
2. There is no hint in the documentation that sorted labels are generally expected and unsorted labels break staleness handling.

**Comment by evkuzin** [comment]
Yeah, sorry, I didn't search for similar issues and just opened a new one. 
My issue is similar, but the staleness comes from vmagent, which does the same, basically in the same GetRowsDiff.
I understand that on the one hand, such behaviour for staleness is useful, but on the other hand, it is a bug because there is no mention in any documents that the client should emit sorted labels. Although Prometheus client does [do that ](https://github.com/prometheus/client_golang/blob/main/prometheus/value.go#L235) 
workaround for no_stale_markers is also not very good, because it is a setting per job, but what if my scrape file contains something like this:
```
    - source_labels: [ __meta_consul_service ]
      regex: (\w+)[_-]exporter
      target_label: job
      replacement: $1
```
Basically, I have to extract this specific job to a separate file if I want a workaround for a particular job, right? Not great design IMO.

**Comment by hagen1778** [comment]
One more workaround for stale markers produced due to inconcsistent order of label pairs between scrapes is to set `-dedup.minScrapeInterval=1ms` on vmselect and vmstorage. That would remove duplicates with the same timestamp and will prefer numerical value to stale marker on conflicts.

This change will have retro-active effect - displaying of historical data will be fixed too.
This change can be applied on vmselect only - then original data in vmstorage will remain unmodified.
